### ISSUE TITLE
electron-builder 21 feedback

### ISSUE THREAD

**Issue by develar** [electron-userland/electron-builder#4018]
Please try version 21. Please do not file issues about old version (< 21), quite a lot of things were fixed and polished in 21.

21 is ready for production use, but it is not marked as "release" yet, because to simplify [your possible migration](https://github.com/electron-userland/electron-builder/releases/tag/v21.0.1) (or addition) from deb/AppImage to snap, [support publishing to Snapcraft](https://github.com/electron-userland/electron-builder/issues/3187) is a blocking issue for release.

electron-updater >= 4.11.0 is recommended to use with electron-builder 21.

**Comment by ZaneL** [comment]
I haven't gotten a chance to try the new version yet on my Pi...hopefully I can in the next day or two. I think it would be really neat to get electron snaps working on Ubuntu Core on a raspberry pi.

To get graphical snaps working on Ubuntu Core, you have to install mir-kiosk https://snapcraft.io/mir-kiosk which acts as the display server. Then you have to integrate a plugin called "xwayland-kiosk-helper" into the electron snap which converts electron's X11 calls into wayland calls. The electron snap would then communicate with mir-kiosk to display the content.

https://tutorials.ubuntu.com/tutorial/electron-kiosk#3

**Comment by develar** [comment]
@ZaneL Thanks, now I understand how do I can install ubuntu core and for what template for arm64 should be provided.

**Comment by ZaneL** [comment]
I would just install this image that is ready to go:

https://ubuntu.com/download/iot/raspberry-pi-2-3-core

It is a standard armhf 32 bit image. It's basically just a super stripped down linux kernel with snapd installed, so that you can install snaps and run them.

**Comment by sapkra** [comment]
electron-builder 21 was released (21.0.15) but there are no updated docker images.
When are you planning the upload them to Docker Hub?

**Comment by xiaoxiangmoe** [comment]
The electron version is 4.0.8 and the electron-builder version is 21.0.5.

After the compilation is complete and installing app on another brand new and clean mac system, the electron app crashes directly when it comes to obtaining permissions such as camera or voice.

I solved the problem by downgrading to electron-builder 20.

**Comment by audiolion** [comment]
I seem to be having an issue where file macros arent being expanded, when publishing to s3 with the following config:

```
"publish": {
      "provider": "s3",
      "region": "us-east-1",
      "bucket": "cdn.kimmel.io",
      "channel": "${env.NEXUS_ELECTRON_CHANNEL}",
      "path": "${env.NEXUS_ELECTRON_S3_PATH}",
      "acl": "public-read"
    }
```

the app was published to:

![image](https://user-images.githubusercontent.com/2430381/63448847-d98c1a00-c40c-11e9-960c-b92399b042e0.png)

it interpreted the path literally. 

I am using an `electron-builder.env` file in the current directory I am running the cli command from. that doesnt seem to be working because the publish failed until I moved the `AWS_ACCESS_KEY_ID` and `AWS_ACCESS_SECRET_KEY` to `~/.aws/credentials`.

I tried exporting from the command line before running, same result, uploaded to the literal path and didnt expand the macro. I also tried exporting within the command itself e.g.

```
export NEXUS_ELECTRON_S3_PATH=/desktop && yarn build:desktop
```

and still the same result.

**Comment by Topppy** [comment]
the latest electron-updater version I can get on npm is 4.1.2. How can  I get electron-updater >= 4.11.0?

**Comment by danwid** [comment]
I'm having similar issues with one of the referred issues 
https://github.com/electron-userland/electron-builder/issues/3684

`Cannot download differentially, fallback to full download: Error: Maximum allowed size is 5 MB`
>>windows 8.1 pro
private github repo
electron-builder 21.2.0
electron 6.0.7
electron-updater 4.1.2

**Comment by wesionaire** [comment]
> I'm having similar issues with one of the referred issues
> #3684
> 
> `Cannot download differentially, fallback to full download: Error: Maximum allowed size is 5 MB`

We have the same problem (private Github repo)

**Comment by mcous** [comment]
Hey @develar, thanks for all your hard work on this project! I'm in the process of updating our [Electron app](https://github.com/Opentrons/opentrons/tree/edge/app-shell) from electron-builder v20 to v21, and I wanted to share my feedback in case it's helpful

- electron-builder: 21.2.0
- electron-updater: 4.1.2
- electron: 6.0.7
- targets: macOS (DMG, ZIP), Windows (NSIS), Linux (AppImage)

TL;DR: I feel like I need to upgrade to v21 for my macOS users and the imminent Catalina release with notarization requirements, but:

- I'm unable to upgrade to the latest v21 on Windows because of a registry bug
- My (new) Linux users are going to lose desktop integration because AppImageLauncher is not compatible with electron-updater and doesn't seem to be a fully working product.

### macOS

The upgrade for macOS has gone mostly smoothly. Notarization for macOS 10.15 Catalina with the DMG target works really well!

The ZIP target has weird issues with macOS's built-in Archive Utility - See https://github.com/electron-userland/electron-builder/issues/4179#issuecomment-528575675 for a detailed breakdown.

### Windows

We've had to lock electron-builder to 21.1.2 on Windows due to a breaking change to Windows uninstall registry entries introduced by #4069 and landed in 21.1.3.

Due to the change in the registry key format of the uninstall registry entry and the failure of the installer to remove for old-format keys, the user is left with duplicate entries in their "Add and Remove Programs" utility after upgrading their app. See #4092 for more details.

### Linux

IMO the removal of built-in Desktop integration in favor of AppImageLauncher was premature. In my testing, AppImageLauncher is both not ready for prime time nor does it interact well with electron-update. I have encountered the following issues:

- Intermittent missing files in AppImages copied to the `Applications` directory by AppImageLauncher
    - In the worst case, I ended up with AppImages that were missing `app-update.yml`!
    - I haven't created a minimal repro for this, but I can if you're interested
- AppImageLauncher is *completely incompatible* with electron-updater's AppImage update mechanism
    - electron-updater relies on `process.env.APPIMAGE` to be the location of the source AppImage to replace
    - AppImageLauncher launches the AppImages from an AppImageLauncher-specific directory, resulting the `process.env.APPIMAGE` to be set to something like `/run/user/1000/appimagelauncherfs/0001.AppImage`
    - electron-updater happens to error out when it attempts to `unlinkSync` that file, but even if it were able to remove it, it would try to move the upgrade file to the completely wrong place rather than `~/Applications`, where AppImageLauncher keeps the source AppImages
    - See #4046 for more details

**Comment by TanninOne** [comment]
#3185 is still not fixed (as of 21.2.0) so stuck with 20.15.0

**Comment by AleksMeshkov** [comment]
[4092](https://github.com/electron-userland/electron-builder/issues/4092) is also doesn't seem to be fixed.

**Comment by quanglam2807** [comment]
Is there a temporary remedy for this issue?

**Comment by keithslater** [comment]
Is this any solution for this? Not sure why this issue is closed since it's still broken.

**Comment by TheAssassin** [comment]
Regarding AppImageLauncher, please see https://github.com/electron-userland/electron-builder/issues/4046#issuecomment-699743169.
